app/src/main/java/com/keylesspalace/tusky/components/timeline/viewmodel/CachedTimelineViewModel.kt

```diff
@@ -0,0 +1,187 @@
+import com.keylesspalace.tusky.network.TimelineCases
+import com.keylesspalace.tusky.util.dec
+import com.keylesspalace.tusky.util.inc
+import com.keylesspalace.tusky.viewdata.StatusViewData
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.rx3.await
+import javax.inject.Inject
+
+class CachedTimelineViewModel @Inject constructor(
+    timelineCases: TimelineCases,
+    private val api: MastodonApi,
+    eventHub: EventHub,
+    accountManager: AccountManager,
+    sharedPreferences: SharedPreferences,
+    filterModel: FilterModel,
+    private val db: AppDatabase,
+    private val gson: Gson
+) : TimelineViewModel(timelineCases, api, eventHub, accountManager, sharedPreferences, filterModel) {
+
+    @ExperimentalPagingApi
+    override val statuses = Pager(
+        config = PagingConfig(pageSize = LOAD_AT_ONCE),
+        remoteMediator = CachedTimelineRemoteMediator(accountManager, api, db, gson),
+        pagingSourceFactory = { db.timelineDao().getStatusesForAccount(accountManager.activeAccount!!.id) }
+    ).flow
+        .map { pagingData ->
+            pagingData.map { timelineStatus ->
+                timelineStatus.toViewData(gson)
+            }
+        }
+        .map { pagingData ->
+            pagingData.filter { statusViewData ->
+                !shouldFilterStatus(statusViewData)
+            }
+        }
+        .cachedIn(viewModelScope)
+
+    override fun updatePoll(newPoll: Poll, status: StatusViewData.Concrete) {
+        // handled by CacheUpdater
+    }
+
+    override fun changeExpanded(expanded: Boolean, status: StatusViewData.Concrete) {
+        viewModelScope.launch {
+            db.timelineDao().setExpanded(accountManager.activeAccount!!.id, status.id, expanded)
+        }
+    }
+
+    override fun changeContentHidden(isShowing: Boolean, status: StatusViewData.Concrete) {
+        viewModelScope.launch {
+            db.timelineDao().setContentHidden(accountManager.activeAccount!!.id, status.id, isShowing)
+        }
+    }
+
+    override fun changeContentCollapsed(isCollapsed: Boolean, status: StatusViewData.Concrete) {
+        viewModelScope.launch {
+            db.timelineDao().setContentCollapsed(accountManager.activeAccount!!.id, status.id, isCollapsed)
+        }
+    }
+
+    override fun removeAllByAccountId(accountId: String) {
+        viewModelScope.launch {
+            db.timelineDao().removeAllByUser(accountManager.activeAccount!!.id, accountId)
+        }
+    }
+
+    override fun removeAllByInstance(instance: String) {
+        viewModelScope.launch {
+            db.timelineDao().deleteAllFromInstance(accountManager.activeAccount!!.id, instance)
+        }
+    }
+
+    override fun removeStatusWithId(id: String) {
+        // handled by CacheUpdater
+    }
+
+    override fun loadMore(placeholderId: String) {
+        viewModelScope.launch {
+            val response = api.homeTimeline(maxId = placeholderId.inc(), limit = 20).await()
```

Comment:
we check response in other places but not here

Reply:
thx